src/Http/Routing/src/Builder/EndpointRouteBuilderExtensions.cs

```diff
@@ -196,7 +196,23 @@ public static RouteGroupBuilder MapGroup(this IEndpointRouteBuilder endpoints, R
         ArgumentNullException.ThrowIfNull(pattern);
         ArgumentNullException.ThrowIfNull(requestDelegate);
 
-        return endpoints.GetOrAddRouteEndpointDataSource().AddRequestDelegate(pattern, requestDelegate, httpMethods);
+        return endpoints
+            .GetOrAddRouteEndpointDataSource()
+            .AddRequestDelegate(pattern, requestDelegate, httpMethods, CreateHandlerRequestDelegate);
+
+        static RequestDelegateResult CreateHandlerRequestDelegate(Delegate handler, RequestDelegateFactoryOptions options, RequestDelegateMetadataResult? metadataResult)
+        {
+            var requestDelegate = (RequestDelegate)handler;
+
+            // Create request delegate that calls filter pipeline.
+            if (options.EndpointBuilder?.FilterFactories.Count > 0)
+            {
+                requestDelegate = RequestDelegateFilterPipelineBuilder.Create(requestDelegate, options);
+            }
+
+            var metadata = new List<object>(options.EndpointBuilder?.Metadata ?? Array.Empty<object>());
```

Comment:
```suggestion
            IReadOnlyList<object> metadata = options.EndpointBuilder?.Metadata is not null ?
                new List<object>(options.EndpointBuilder?.Metadata) :
                Array.Empty<object>();
```

Shouldn't need to allocate an empty list here.